pkg/collector/stats/types/types.go

```diff
@@ -38,90 +46,92 @@ func (s *UInt64Stat) SetNewDeltaValue(newDelta uint64, sum bool) error {
 		return nil
 	}
 	// verify overflow
-	if s.Aggr >= (math.MaxUint64 - newDelta) {
+	if s.aggr.Load() >= (math.MaxUint64 - newDelta) {
 		// we must set the the value to 0 when overflow, so that prometheus will handle it
-		s.Aggr = 0
+		s.aggr.Store(0)
 		return fmt.Errorf("the aggregated value has overflowed")
 	}
 	if sum {
 		// sum is used to accumulate metrics from different processes
-		s.Delta += newDelta
+		s.delta.Add(newDelta)
 	} else {
-		s.Delta = newDelta
+		if swapped := s.delta.CompareAndSwap(s.delta.Load(), newDelta); !swapped {
+			return fmt.Errorf("the delta value has been changed by another process")
+		}
 	}
-	s.Aggr += newDelta
+	// update aggregated value
+	s.aggr.Add(newDelta)
 	return nil
 }
 
 // SetNewAggr set new read aggregated value (e.g., from cgroup, energy files)
 func (s *UInt64Stat) SetNewAggr(newAggr uint64) error {
-	if newAggr == 0 || newAggr == s.Aggr {
+	currAggr := s.aggr.Load()
+	if newAggr == 0 || newAggr == currAggr {
 		// if a counter has not changed, we skip it
 		return nil
 	}
 	// verify aggregated value overflow
 	if newAggr == math.MaxUint64 {
 		// we must set the the value to 0 when overflow, so that prometheus will handle it
-		s.Aggr = 0
+		s.aggr.Store(0)
```

Comment:
shouldn't we also reset `delta`?

Reply:
I'm not entirely sure we should be resetting at all.
I know the expectation from prometheus's side is that counters get reset to 0 on restart.
I would think it makes sense to pick the right precision of counter in the first place 🤔

Reply:
Note necessarily reset to 0 because it can't scrape exactly when the reset to 0 happened. 
AFAIK, the expectation from Prometheus is that counters are monotonic and are ever increasing, so if the previous scrape is greater than current then, `rate | sum| etc..` would assume there was a reset in between. 
see: https://www.robustperception.io/how-does-a-prometheus-counter-work/#:~:text=A%20common%20question%20is%20what%20happens%20when%20the%20process%20restarts%20and%20the%20counter%20is%20reset%20to%200%3F%20rate()%C2%A0will%C2%A0automatically%20handle%20this.%20Any%20time%20a%20counter%20appears%20to%20decrease%20it%27ll%20be%20treated%20as%20though%20there%20was%20a%20reset%20to%200%20right%20after%20the%20first%20data%20point.